@ndlib/gatsby-theme-marble/src/components/App/Pages/User/UserBody/UserEdit/index.js

```diff
@@ -0,0 +1,146 @@
+  const claims = typy(loginReducer, 'token.claims').safeObject
+  const [fullName, changeName] = useState(user.fullName)
+  const [email, changeEmail] = useState(user.email)
+  const [bio, changeBio] = useState(user.bio)
+  const [patching, setPatching] = useState(false)
+
+  if (!ownsPage(loginReducer, user.userName)) {
+    return (<Unauthorized />)
+  }
+
+  return (
+    <form className={style.edit}>
+      <div className={style.buttonGroup}>
+        <MaterialButton
+          onClick={(e) => {
+            e.preventDefault()
+            if (window.confirm(
+              `Any unsaved changes you have made will be lost.`,
+            )) {
+              navigate(`/user/${user.userName}`)
+            }
+          }}
+          disabled={patching}
+        >Cancel</MaterialButton>
+        <MaterialButton
+          onClick={(e) => {
+            e.preventDefault()
+            setPatching(true)
+            const body = {
+              fullName: fullName,
+              email: email,
+              bio: bio || null,
+              uuid: `${claims.sub}.${btoa(claims.iss)}`,
+              userName: claims.netid,
+            }
+            patchData(loginReducer, body)
+          }}
+          disabled={patching}
+          primary
+        >Save</MaterialButton>
+      </div>
+      <TextField
+        id='profileName'
+        label='Name'
+        defaultValue={user.fullName}
+        onChange={(event) => {
+          changeName(event.target.value)
+        }}
+        disabled={patching}
+      />
+      <TextField
+        id='profileUserName'
+        label='Username'
+        defaultValue={user.userName}
+        disabled
+      />
+      <TextField
+        id='profileEmail'
+        label='Email Address'
+        defaultValue={user.email}
+        onChange={(event) => {
+          changeEmail(event.target.value)
+        }}
+        disabled={patching}
+      />
+      <TextArea
+        id='profileBio'
+        label='Bio'
+        defaultValue={user.bio}
+        onChange={(event) => {
+          changeBio(event.target.value)
+        }}
+        disabled={patching}
+      />
+      <div>
+        <label
+          htmlFor='profileGravatar'
+          className={style.editLabel}
+        >Avatar</label>
+        <div
+          id='profileGravatar'
+          className={style.gravatarEdit}
+        >
+          <Gravatar email={user.email} size={100} />
+          <span>User icons are provided by <a href='https://en.gravatar.com'>Gravatar</a>, the globally recognized avatar service.</span>
+        </div>
+      </div>
+    </form>
+  )
+}
+
+UserEdit.propTypes = {
+  user: PropTypes.object.isRequired,
+  loginReducer: PropTypes.object.isRequired,
+}
+
+export const mapStateToProps = (state) => {
+  return { ...state }
+}
+const mapDispatchToProps = dispatch => {
+  return { dispatch }
+}
+export default connect(
+  mapStateToProps,
+  mapDispatchToProps,
+)(UserEdit)
+
+const patchData = (loginReducer, body) => {
```

Comment:
Maybe this would be appropriate to put in actions or a separate file? Also that TODO looks like it might be outdated.

Reply:
I don't really think actions is the appropriate place to put this, since it doesn't interact with the reducer, but I agree this may not be the best place for this logic. It will probably move depending on how similar updating collections ends up being as this may be able to become a generic function shared for all form updates.